sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -180,6 +180,56 @@ static inline std::string GetHTTPMessagePreBody(Azure::Core::Http::Request const
 
   return httpRequest;
 }
+
+static void CleanupThread()
+{
+  using namespace Azure::Core::Http::_detail;
+  for (;;)
+  {
+
+    // Won't continue until the ConnectionPoolMutex is released from MoveConnectionBackToPool
+    std::unique_lock<std::mutex> lockForPoolCleaning(
+        CurlConnectionPool::g_curlConnectionPool.ConnectionPoolMutex);
+
+    // Wait the defined default time OR to the signal from the conditional variable.
+    // wait_for releases the mutex lock until it wakes up again or it's cancelled.
+    if (CurlConnectionPool::g_curlConnectionPool.ConditionalVariableForCleanThread.wait_for(
+            lockForPoolCleaning,
+            std::chrono::milliseconds(DefaultCleanerIntervalMilliseconds),
+            []() {
+              return CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.size() == 0;
+            }))
+    {
+      // Cancelled by another thead or no connections on wakeup
+      return;
+    }
+
+    // loop the connection pool index - Note: lock is re-taken for the mutex
+    // Notes:
+    // - If all connections are removed from a host-index, the index is not removed from the pool to
+    // reduce the time of the entire routine.
```

Comment:
I don't understand what you mean here? "the index is not removed" By index, do you mean that the map key doesn't get removed? Is that actually true, and if so, why?

Reply:
Yes, the entry in the dictionary of <string, list<connection>> is not removed by the clean routine.

This is because the clean thread, loops the dictionary for each index-entry. Removing one index-entry while looping requires to adjust the way the loop is done.  I think I can update it to keep it consistant.